pkg/app/piped/imagewatcher/watcher.go

```diff
@@ -13,68 +13,179 @@
 // limitations under the License.
 
 // Package imagewatcher provides a piped component
-// that periodically checks the image registry and updates
+// that periodically checks the container registry and updates
 // the image if there are differences with Git.
 package imagewatcher
 
 import (
 	"context"
+	"sync"
 	"time"
 
 	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/app/piped/imageprovider"
+	"github.com/pipe-cd/pipe/pkg/config"
+	"github.com/pipe-cd/pipe/pkg/git"
 )
 
 type Watcher interface {
 	Run(context.Context) error
 }
 
-type watcher struct {
-	timer  *time.Timer
-	logger *zap.Logger
+type gitClient interface {
+	Clone(ctx context.Context, repoID, remote, branch, destination string) (git.Repo, error)
 }
 
-type imageRepos map[string]imageRepo
-type imageRepo struct {
+type watcher struct {
+	config    *config.PipedSpec
+	gitClient gitClient
+	logger    *zap.Logger
+
+	// Indexed by repo id.
+	gitRepos sync.Map
 }
 
-func NewWatcher(interval time.Duration, logger *zap.Logger) Watcher {
+func NewWatcher(cfg *config.PipedSpec, gitClient gitClient, logger *zap.Logger) Watcher {
 	return &watcher{
-		timer:  time.NewTimer(interval),
-		logger: logger,
+		config:    cfg,
+		gitClient: gitClient,
+		logger:    logger.Named("image-watcher"),
 	}
 }
 
+// Run spawns goroutines for each image provider.
 func (w *watcher) Run(ctx context.Context) error {
+	// Pre-clone to cache the registered git repositories.
+	for _, r := range w.config.Repositories {
+		repo, err := w.gitClient.Clone(ctx, r.RepoID, r.Remote, r.Branch, "")
+		if err != nil {
+			w.logger.Error("failed to clone repository",
+				zap.String("repo-id", r.RepoID),
+				zap.Error(err),
+			)
+			return err
+		}
+		w.gitRepos.Store(r.RepoID, repo)
+	}
+
+	for _, cfg := range w.config.ImageProviders {
+		p, err := imageprovider.NewProvider(&cfg, w.logger)
+		if err != nil {
+			return err
+		}
+
+		go w.run(ctx, p, cfg.PullInterval.Duration())
+	}
+	return nil
```

Comment:
This is exited immediately after starting `goroutine`s for all image providers.
Please ensure that all of its child tasks must be stopped before exiting.